### ISSUE TITLE
Open H.264 Missing Colour Space

### ISSUE THREAD

**Issue by any1** [rfbproto/rfbproto#60]
The colour space with which frames are encoded is not included in the H.264 bit stream. It is also not included anywhere in Open H.264.

We should try to rectify this. Otherwise, colours will be slightly off in many cases when using H.264 encoding. It should at least be specified in the protocol which colour space is to be used. I'm not sure though if all encoders are able to apply different colour spaces, so we might need to create new modified extension to include them, for good measure.

**Comment by CendioOssman** [comment]
Do you know which color spaces the current implementations are assuming?

**Comment by any1** [comment]
> Do you know which color spaces the current implementations are assuming?

No, it's all over the place.

**Comment by any1** [comment]
Actually, I found this under section E.2.1: VUI Parameter Semantics

It was just a matter of reading up on colour spaces and grepping for other words.

I was also a little bit confused about why there would be an API for setting colour space in the WebCodecs API when you would expect something like this to be coded into the bitstream. It turns out that the reason for that is that these things can be set as "unspecified" in the bitstream, in which case the colour space is expected to be handled out-of-band.

In conclusion: Servers should make sure to encode the colour space into the bitstream because it will be unspecified otherwise and the client won't know exactly how to treat it.

We could say that if this information is not given in the stream, clients should default to something specific, but it might be simplest to just specify that this should always be a part of the stream.

**Comment by CendioOssman** [comment]
Sounds reasonable. Assuming that we aren't already in a situation where most servers leave things unspecified. :)

I don't suppose there is some de facto norm for other users of H.264 that we could use? E.g., what happens when ffmpeg decodes a file that lacks those parameters?

**Comment by any1** [comment]
> I don't suppose there is some de facto norm for other users of H.264 that we could use? E.g., what happens when ffmpeg decodes a file that lacks those parameters?

I think that this is often specified in the container format or the encapsulating streaming protocol.

However, if you play a raw h264 file with mpv, and that file doesn't have colour space info, I think that mpv has no other choice than to pick a default one. FFmpeg doesn't assign one if it's unspecified in the protocol, it just tells the player that it is unspecified, so it's up to whatever renders the video.

**Comment by any1** [comment]
It looks like PiKVM's @mdevaev is already working on the problem: https://github.com/raspberrypi/firmware/issues/1885

**Comment by mdevaev** [comment]
Let's make the behavior so that it matches the players as described [here](https://github.com/raspberrypi/firmware/issues/1885#issuecomment-2294799301)

> In the absence of explicit headers in the bitstream, players will generally assume limited range.

So for #64 I suggest adding that it is recommended that the servers specify the listed headers in the bit stream, otherwise a limited range is assumed. It should make life easier for both the client and the server, plus it won't be any special behavior, since it's common in the H.264 world.

**Comment by CendioOssman** [comment]
Which limited range, though? BT.601? 709?

**Comment by any1** [comment]
> Which limited range, though? BT.601? 709?

According to the Raspberry Pi developer, it should be BT.601 for SD content and BT.709 for HD content. I'd argue that we can't really be sure of anything in the client if the colour space is not encoded into the stream.

I'm not sure if putting extra logic for this into the client is necessarily a good idea. It's probably just as good to rely on the decoder's or the video player's defaults. If the colours are wrong, it's the server's fault anyway.

**Comment by any1** [comment]
There is at least one other thing in the bit stream that can be left unspecified by the encoder.

We might as well also say that if `aspect_ratio_idc` is unspecified, it should be treated as 1:1.

The other parameters that can be unspecified appear to not be of any significance.

**Comment by mdevaev** [comment]
> I'm not sure if putting extra logic for this into the client is necessarily a good idea. It's probably just as good to rely on the decoder's or the video player's defaults. If the colours are wrong, it's the server's fault anyway.

Then perhaps we should require specifying the color space explicitly. If not specified, then the client's decoder is free to act at its discretion, but to display at least something, even with the wrong colors, than nothing at all. This should be a common behavior: something is better than nothing.

> Edit: Perhaps, we should just say that it must always be 1:1. Otherwise clients would have to implement funky pixel aspect ratios.

Agree.

> By the way, sei.display_orientation would be good for optimisation purposes, but it would be better to implement as a generic VNC extension.

Agree.

**Comment by CendioOssman** [comment]
So is #62 good to go? Or should be adjusted based on the discussion here?

**Comment by any1** [comment]
> So is #62 good to go? Or should be adjusted based on the discussion here?

I was thinking of adding some more clarifications about which parameters need to be set in the stream and which parameters are to be disregarded. However, #62 is good enough on its own and I don't foresee revisiting this soon, so I suppose that we might as well just merge it.
